src/main/java/com/microsoft/sqlserver/jdbc/KeyVaultCredential.java

```diff
@@ -24,20 +24,30 @@
  */
 class KeyVaultCredential extends KeyVaultCredentials {
 
+	SQLServerKeyVaultAuthenticationCallback authenticationCallback = null;
     String clientId = null;
     String clientKey = null;
+    String accessToken = null;
 
     KeyVaultCredential(String clientId,
             String clientKey) {
         this.clientId = clientId;
         this.clientKey = clientKey;
     }
-
+    
+    KeyVaultCredential(SQLServerKeyVaultAuthenticationCallback authenticationCallback) {
+        this.authenticationCallback = authenticationCallback;
+    }
+    
     public String doAuthenticate(String authorization,
             String resource,
             String scope) {
-        AuthenticationResult token = getAccessTokenFromClientCredentials(authorization, resource, clientId, clientKey);
-        return token.getAccessToken();
+    	if(authenticationCallback==null) {
```

Comment:
Maybe we could try to avoid multiple returns?

https://github.com/Microsoft/mssql-jdbc/blob/dev/Coding_Guidelines.md